django/contrib/auth/password_validation.py

```diff
@@ -172,12 +172,10 @@ class CommonPasswordValidator:
     def __init__(self, password_list_path=DEFAULT_PASSWORD_LIST_PATH):
         try:
             with gzip.open(str(password_list_path)) as f:
-                common_passwords_lines = f.read().decode().splitlines()
+                self.passwords = set(map(lambda x: x.decode().strip(), f))
```

Comment:
I think `{line.decode().strip() for line in f}` reads a little better. Also for the one below.

BTW, since Django no longer supports Python 2.7, you can remove the explicit `decode()` step by passing `encoding='utf-8'` to `gzip.open()`, then you'll get a text-mode file rather than a binary-mode one.

Finally, if you want to keep the previous idea of not duplicating the set/strip logic, you can write it like this:

```python
try:
   f = gzip.open(str(password_list_path), encoding='utf-8')
except OSError:
    f = open(str(password_list_path))
with f:
    self.passwords = {line.strip() for line in f}
```

Reply:
But the default `mode` for `gzip.open` is `'rb'`; doesn't that necessitate decoding the `bytes` content to `str`?

In agreement with your first point about using a set comprehension.

Reply:
Right, sorry, forget about `encoding`, should be `gzip.open(str(password_list_path), mode='rt')` according to [the doc](https://docs.python.org/3.7/library/gzip.html#gzip.open). (Notice it's `rt`, not the usual `r`).

Reply:
Yep, you've got it. Thanks, will make that change.

Reply:
Hm, your logic on deduplicating the `set` comprehension makes sense also, although if anything besides an `OSError` was raised, then `f` would raise an uncaught `NameError`.

Reply:
If anything is raised besides `OSError`, it will propagate, it won't continue to the `with` part. This is the same as the existing behavior.

Reply:
Yup, you're right. Duh on my part.

Reply:
One last thing - regarding `encoding` - your thinking there is that we want to be explicit about utf-8 rather than deferring to `locale.getpreferredencoding(False)` as ultimately otherwise gets called in `io.TextIOWrapper`?

Reply:
Huh, I assumed that the default is `utf-8`, but seems like it's not. So yes, should pass `encoding='utf-8'` to keep the previous behavior (for `bytes.decode()`, 'utf-8' is the default).

Reply:
Alright, it looks like this is now failing:

```
======================================================================
ERROR [0.002s]: test_validate_custom_list (auth_tests.test_validators.CommonPasswordValidatorTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.7/tests/auth_tests/test_validators.py", line 196, in test_validate_custom_list
    validator = CommonPasswordValidator(password_list_path=path)
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.7/django/contrib/auth/password_validation.py", line 178, in __init__
    self.passwords = {x.strip() for x in f}
  File "/home/jenkins/workspace/pull-requests-bionic/database/sqlite3/label/bionic-pr/python/python3.7/django/contrib/auth/password_validation.py", line 178, in <setcomp>
    self.passwords = {x.strip() for x in f}
  File "/usr/lib/python3.7/gzip.py", line 289, in read1
    return self._buffer.read1(size)
  File "/usr/lib/python3.7/_compression.py", line 68, in readinto
    data = self.read(len(byte_view))
  File "/usr/lib/python3.7/gzip.py", line 463, in read
    if not self._read_gzip_header():
  File "/usr/lib/python3.7/gzip.py", line 411, in _read_gzip_header
    raise OSError('Not a gzipped file (%r)' % magic)
OSError: Not a gzipped file (b'fr')
```

Failing test is here:

https://github.com/django/django/blob/cbf7e71558c94ce1c327b683768a18a25d82d197/tests/auth_tests/test_validators.py#L194

No idea why the `OSError` isn't caught like it should be though.  This is the point of the `except` to read a plain text file such as `common-passwords-custom.txt` as in the test.

Any idea @bluetech

Reply:
(hard-resetting the recent encoding commit to give that a shot...)

Reply:
From the stacktrace you pasted, it looks like the problem was that we assumed that it is `gzip.open()` that would throw `OSError` if the file is not a gzip, but in fact it only happens when the file is first read. So, it's not the `encoding` that's the problem (this should stay).

So I think this is what we want - this time I did try it out :)

```python
try:
    with gzip.open('test', mode='rt', encoding='utf-8') as f:
        self.passwords = {password.strip() for password in f}        
except OSError:
    with open('test') as f:
        self.passwords = {password.strip() for password in f}
```

Reply:
Aha - you're right again.  Confirmed with:

```bash
$ echo "not a gzip file!" > notgz.txt.gz
$ python3 -c "import gzip; f = gzip.open('notgz.txt.gz', 'rt')"  # OK
$ python3 -c "import gzip; f = gzip.open('notgz.txt.gz', 'rt'); list(f)"  # Raises
```

Will update when I have a second this evening.  Thanks for your help here @bluetech.